Move IFixedStatement to internal

333fred opened the pull request:
It's likely that IFixedStatement will be fine as is, but we're cutting it for V1. So, replace it with NoneOperation publicly. As part of this, I had to remove NoneOperation filtering from IBlockStatement. In C# it's entirely gone, in VB I'm filtering method/accessor/operator declarations, as those appear in the underlying BoundBlock. Tagging @dotnet/analyzer-ioperation @dotnet/roslyn-compiler for review.

cston commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -997,7 +998,7 @@ private IParameterInitializer CreateBoundParameterEqualsValueOperation(BoundPara
         {
             Lazy<ImmutableArray<IOperation>> statements =
-                new Lazy<ImmutableArray<IOperation>>(() => boundBlock.Statements.Select(s => Create(s)).Where(s => s.Kind != OperationKind.None).ToImmutableArray());
+                new Lazy<ImmutableArray<IOperation>>(() => boundBlock.Statements.Select(s => Create(s)).ToImmutableArray());
```
`boundBlock.Statements.SelectAsArray(s => Create(s))`

cston commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -927,7 +927,16 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                   statement.Syntax.Kind() <> SyntaxKind.GetAccessorStatement AndAlso
+                                   statement.Syntax.Kind() <> SyntaxKind.SetAccessorStatement
+                        End Function
+                    ).ToImmutableArray()
```
Switch `Where` and `Select` to avoid creating instances that are dropped.

cston replied:
Are there other statement kinds that should be ignored? `SubNewStatement`, `AddHandlerAccessorStatement`, ...?

333fred replied:
`SubNewStatement` already has a test. I'll add tests for the 3 event blocks as well.

333fred replied:
Actually, those already have tests too. https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb#L44, https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb#L178, https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb#L218, https://github.com/dotnet/roslyn/blob/features/ioperation/src/Compilers/VisualBasic/Test/Semantic/IOperation/IOperationTests_IBlockStatement_MethodBlocks.vb#L258.

333fred commented:
@cston addressed feedback.

AlekseyTs commented on src/Compilers/VisualBasic/Portable/Operations/VisualBasicOperationFactory.vb:
```diff
@@ -982,7 +982,15 @@ Namespace Microsoft.CodeAnalysis.Semantics
+                                Return statement.Syntax.Kind() <> SyntaxKind.OperatorStatement AndAlso
+                                       statement.Syntax.Kind() <> SyntaxKind.FunctionStatement AndAlso
+                                       statement.Syntax.Kind() <> SyntaxKind.GetAccessorStatement AndAlso
+                                       statement.Syntax.Kind() <> SyntaxKind.SetAccessorStatement
```
>statement.Syntax.Kind() <> SyntaxKind.SetAccessorStatement [](start = 39, length = 58)

What about AddHandlerAccessorStatement, RemoveHandlerAccessorStatement and RaiseEventAccessorStatement? It feels like we cannot really be sure what impact this change is going to have. What are we trying to accomplish and how is it related to the ```fixed``` statement?

333fred replied:
The issue here is that, when we move `fixed` statements to be NoneOperation, they are then not included in any block statements, which means that analyzers will miss both the fixed statement and any children it declares. This is removing the filtering, but at the same time that introduces these first-line bound nodes are included in the block. We have tests for the 3 event blocks, and those were unaffected by this change.

AlekseyTs replied:
> which means that analyzers will miss both the fixed statement and any children it declares.

I believe VB does not have ```fixed``` statement.

333fred replied:
It doesn't, but in removing the NoneOperation filter from C# I removed it from VB as well. I haven't gone through the list of outstanding statement types yet to see if this affects any of others in VB, but I'll try to do so by discussion Thursday.

AlekseyTs commented on src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:
```diff
@@ -1057,7 +1058,7 @@ private IParameterInitializer CreateBoundParameterEqualsValueOperation(BoundPara
         private IBlockStatement CreateBoundBlockOperation(BoundBlock boundBlock)
         {
             Lazy<ImmutableArray<IOperation>> statements =
-                new Lazy<ImmutableArray<IOperation>>(() => boundBlock.Statements.Select(s => Create(s)).Where(s => s.Kind != OperationKind.None).ToImmutableArray());
```
>Select(s => Create(s)).Where(s => s.Kind != OperationKind.None).ToImmutableArray()); [](start = 81, length = 84)

I do not feel comfortable about this change, it is likely to have unintended consequences. If the goal is to keep ```fixed``` statement in the list, we should do exactly that, I believe.

AlekseyTs commented:
None of the existing tests required adjustments. Does it mean that we don't have any tests for ```fixed``` statements, or does it mean we don't assert IOperation.Kind in them?

AlekseyTs commented:
Done with review pass.

333fred commented:
@AlekseyTs we don't have tests for `fixed` at the moment. I believe that @mavasani and @CyrusNajmabadi  determined that since we aren't planning on supporting a lot of native features for V1, it would be ok to convert this to NoneOperation and skip for V1. I've added this to the discussion list so we can talk about it on Thursday.

nguerrera commented:
@nguerrera (Please ignore this, checking if explicitly tagging me unmutes my notifications. Sorry for the noise.)

AlekseyTs commented:
> we don't have tests for fixed at the moment.

I think it would be good to have at least one test to assert that we get expected kind and children are visible.

333fred commented:
Fixes https://github.com/dotnet/roslyn/issues/21281.

333fred commented:
@AlekseyTs I've updated the PR to remove the changes to the OperationKind.None filtering. I'm now special casing the NoneOperations created from BoundFixedStatements, and no others. I've also added some tests to show that we're seeing fixed statement children correctly.

AlekseyTs commented:
>         private IFixedStatement CreateBoundFixedStatementOperation(BoundFixedStatement boundFixedStatement)

Is this function used after the change?

---
Refers to: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:1258 in 3b5c526. [](commit_id = 3b5c526f3e1b49147d399569446239ad4a28df9c, deletion_comment = False)

333fred commented:
@AlekseyTs because I commented moved IFixedStatement to internal, and didn't entirely remove it, I wanted to keep the existing Creation methods intact.

AlekseyTs commented on src/Compilers/CSharp/Test/Semantic/IOperation/IOperationTests_IFixedStatement.cs:
```diff
@@ -0,0 +1,237 @@
+    {
+        unsafe
+        {
+            /*<bind>*/fixed(int *p = &i)
```
>int *p = &i [](start = 28, length = 11)

It would be good to test a multiple declarator case.

333fred replied:
Added.

333fred commented:
Ping @dotnet/analyzer-ioperation for a review please.

333fred commented:
Ping @dotnet/analyzer-ioperation. I need a review please.

333fred commented:
VS Integration test failure is https://github.com/dotnet/roslyn/issues/21668.

333fred commented:
@dotnet-bot retest windows_release_vs-integration_prtest please
